subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainArtifactResolver.java

```diff
@@ -61,24 +68,47 @@ public void resolveArtifactsWithType(ComponentResolveMetadata component, Artifac
 
     @Nullable
     @Override
-    public ArtifactSet resolveArtifacts(ComponentResolveMetadata component, ConfigurationMetadata configuration, ArtifactTypeRegistry artifactTypeRegistry, ExcludeSpec exclusions, ImmutableAttributes overriddenAttributes) {
+    public ArtifactSet resolveArtifacts(ComponentResolveMetadataForArtifactSelection component, ArtifactTypeRegistry artifactTypeRegistry, ExcludeSpec exclusions, ImmutableAttributes overriddenAttributes) {
         if (component.getSources() == null) {
             // virtual components have no source
             return NO_ARTIFACTS;
         }
-        ModuleComponentRepository sourceRepository = findSourceRepository(component.getSources());
-        return new MetadataSourcedComponentArtifacts().getArtifactsFor(component, configuration, this, sourceRepository.getArtifactCache(), artifactTypeRegistry, exclusions, overriddenAttributes, calculatedValueContainerFactory);
+        return ArtifactSetFactory.createFromVariantMetadata(component.getId(), component.getResolvedVariants(), component.getAttributesSchema(), overriddenAttributes);
     }
 
     @Override
-    public void resolveArtifact(ComponentArtifactMetadata artifact, ModuleSources sources, BuildableArtifactResolveResult result) {
+    public void resolveArtifact(ModuleVersionIdentifier ownerId, ComponentArtifactMetadata artifact, ModuleSources sources, BuildableResolvableArtifactResult result) {
         ModuleComponentRepository sourceRepository = findSourceRepository(sources);
-
-        // First try to resolve the artifacts locally before going remote
-        sourceRepository.getLocalAccess().resolveArtifact(artifact, sources, result);
-        if (!result.hasResult()) {
-            sourceRepository.getRemoteAccess().resolveArtifact(artifact, sources, result);
+        if (artifact.isOptionalArtifact()) {
+            BuildableArtifactResolveResult probe = new DefaultBuildableArtifactResolveResult();
+            // First try to resolve the artifacts locally before going remote
+            sourceRepository.getLocalAccess().resolveArtifact(artifact, sources, probe);
+            if (!probe.hasResult()) {
+                sourceRepository.getRemoteAccess().resolveArtifact(artifact, sources, probe);
+            }
+            if (!probe.isSuccessful()) {
+                result.doesNotExist();
+                return;
+            }
         }
+        CalculatedValue<File> artifactSource = calculatedValueContainerFactory.create(Describables.of(artifact.getId()), new Supplier<File>() {
```

Comment:
Could you use a lambda for the Supplier? I don't think `dependency-management` subproject requires Java 6.